[Frontend] [Core] Add Tensorizer support for V1, LoRA adapter serialization and deserialization

sangstar opened the pull request:
# V1 support, serializing and deserializing LoRA adapters using `tensorizer`
This PR implements Tensorizer support for V1 and allows LoRA adapters to be serialized and deserialized using `tensorizer`. A test is added to confirm this and serializing LoRA files can be done with the same script that can save vLLM models in `examples/tensorize_vllm_model.py`.

## Summary of changes

- Adjusted the CLI args for `.buildkite/test-pipeline.yaml`'s invocation of the `tensorize_vllm_model.py` example script to additionally save and load a LoRA adapter, testing a model generation after deserialization to confirm the LoRA adapter was loaded properly
- Added `--lora-path` to `tensorize_vllm_model.py` as a base argparser argument that allows a user to specify the HuggingFace reference ID for a LoRA adapter, which can be either serialized or deserialized depending on whether the `serialize` or `deserialize` subparser is indicated.
- Added `test_serialize_and_deserialize_lora` to `test_tensorizer.py` testing saving and loading LoRA adapters with `tensorizer`.
- Allows deserializing a LoRA adapter by allowing `TensorizerConfig` to be passed in as a kwarg to `LoRARequest`. When this is done, the LoRA tensors are assumed to be in `tensorizer`'s `.tensors` format and deserialized according to the parameters given in the `TensorizerConfig` provided.
- Added fixes to support V1 for Tensorizer serialization and deserialization, including saving methods for the V1 worker to use with `collective_rpc`. 

FIX #18529


<!--- pyml disable-next-line no-emphasis-as-heading -->

sangstar committed:
feat: Add LoRA adapter support for Tensorizer
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Add partial support for vLLM V1 with Tensorizer
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Update LoRA support from upstream changes
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Rm experimental changes
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Elide type checking issue
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Update snippet, fix tests, fix default `lora_dir`, `tensorizer_uri` behavior
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

jeejeelee reviewed (changes requested):
Please first resolve failures similar to `Undefined name TensorizerConfig`.

sangstar committed:
fix: Fix `TensorizerConfig` undefined issue

`msgspec` was likely eagerly evaluating
the `TensorizerConfig` type annotation,
so explicit import resolution was needed,
defaulting to a `None` dummy value if
the user doesn't have Tensorizer installed.
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Resolve imports for annotations, pre-commit
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Don't run V1 tests for serialization
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Run pre-commit
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Enforce `tensorizer_uri` as only string-typed for linter
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Use assertion to ensure `mypy` is satisfied with `tensorizer_uri`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Switch to `Union` annotation for Python 3.9
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Provide `reason` strings to `skipif` conditions
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Temporarily set `"VLLM_USE_V1" = "0"` in Tensorizer fixture
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar commented:
> Please first resolve failures similar to `Undefined name TensorizerConfig`.

Resolved!

jeejeelee commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -461,8 +473,56 @@ def tensorize_vllm_model(engine_args: EngineArgs,
 
+    if envs.VLLM_USE_V1:
+        raise RuntimeError("Tensorizer serialization requires "
+                           "the env var VLLM_USE_V1=0 set.")
```
QQ: Does `Tensorizer` only support V0?

sangstar replied:
This was initially an implication I had gleaned from this [code snippet](https://github.com/vllm-project/vllm/blob/0189a65a2e76b94ee4cc872dbad2b732256e0e9a/vllm/engine/arg_utils.py#L1186-L1197), but this seems to just be flagged provisionally due to the Tensorizer tests not passing for V1 out-of-the-box. 

My commits I've just pushed in this PR in [2fbd2db](https://github.com/vllm-project/vllm/pull/17926/commits/2fbd2db40af65a68b275452cd8753186f3a7665f) and [7b5dd66](https://github.com/vllm-project/vllm/pull/17926/commits/7b5dd66c74226207587f87b7821165541cb1e407) implement V1 support for Tensorizer, so I've removed the flag preventing it. I'm just currently investigating an issue with `torchdynamo` I discovered during testing, which I suspect it a superficial bug just arising from the setup of my test itself rather than arising from real usage.

sangstar committed:
feat: Implement initial changes to support V1
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

jeejeelee commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -462,7 +474,58 @@ def tensorize_vllm_model(engine_args: EngineArgs,
+    import safetensors
+    from huggingface_hub import snapshot_download
+
+    lora_files = snapshot_download(repo_id=lora_path)
```
`lora_path` may be a local directory, so I think you can use this function: https://github.com/vllm-project/vllm/blob/v0.8.5.post1/vllm/lora/utils.py#L197

jeejeelee commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -462,7 +474,58 @@ def tensorize_vllm_model(engine_args: EngineArgs,
+    # the tensors and config filenames are adapter_model.safetensors and
+    # adapter_config.json respectively, so this logic makes the same
+    # assumption
+    tensor_path = os.path.join(lora_files, "adapter_model.safetensors")
```
LoRA weights are not only in safetensor format,see: https://huggingface.co/jeeejeee/baichuan7b-text2sql-spider/tree/main

sangstar committed:
fix: Pass V1 tests, pass `TensorizerConfig` as dict to `LoRARequest`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Allow different `adapter_model` formats
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Don't use percent format
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Fix `torchdynamo` issue with Tensorizer loading
Replaced the `no_init_or_tensor`context manager used in
`TensorizerAgent._init_model` with `torch.device("meta")` as a workaround
to prevent premature GPU memory allocations without confusing `torchdynamo`
about its patching of `torch.empty`.
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
chore: Update `tensorize_vllm_model.py` docstring
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Use context-manager that implements `no_init_or_tensor` traceably for `torchdynamo`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

jeejeelee commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -348,3 +346,54 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+    ("model_name", "lora_uri"),
+    [("meta-llama/Llama-2-7b-hf", "yard1/llama-2-7b-sql-lora-test"),
+     ("baichuan-inc/Baichuan-7B", "jeeejeee/baichuan7b-text2sql-spider")])
+def test_serialize_and_deserialize_lora(tmp_path, model_name, lora_uri):
```
Can you move this test script  to `tests/lora/`?

jeejeelee commented on tests/entrypoints/openai/test_tensorizer_entrypoint.py:
```diff
@@ -0,0 +1,127 @@
+        completion_tokens=5, prompt_tokens=6, total_tokens=11)
+
+
+def test_confirm_deserialize_and_serve(model_uri, tmp_dir,
```
Isn't this test a bit redundant?

sangstar committed:
fix: Move test to LoRA tests in dedicated file, use smaller model for entrypoint test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Rm redundant test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Remove V0 constraint, clean up `tests/tensorizer_loader`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

simon-mo commented:
Hi @sangstar, thank you for all the support so far. 

CODEOWNERS cannot be done due to its requirements for write access only. However, we do have a workaround! 

We use the `mergify` tool: https://github.com/vllm-project/vllm/blob/main/.github/mergify.yml and it can automatically label and [comment](https://github.com/vllm-project/vllm/blob/main/.github/mergify.yml) based on PR's file changes. You are welcomed to submit a PR modifying it so it will tag you!

sangstar commented:
@simon-mo Sounds great! Just put out a PR for it: https://github.com/vllm-project/vllm/pull/18278

sangstar commented on examples/other/tensorize_vllm_model.py:
```diff
@@ -90,11 +91,27 @@
 loading with tensorizer that are given to `TensorizerConfig`, run:
 
-`python -m examples.other.tensorize_vllm_model deserialize --help`
+`python examples/other/tensorize_vllm_model.py deserialize --help`
```
Yes, thanks for the call out. Fixed.

sangstar commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -462,8 +552,78 @@ def tensorize_vllm_model(engine_args: EngineArgs,
+    # vllm.lora.models.LoRAModel.from_local_checkpoint assumes that
+    # the tensors and config filenames are adapter_model.safetensors and
+    # adapter_config.json respectively, so this logic makes the same
+    # assumption
```
Agreed and removed.

sangstar committed:
fix: Resolve next round of review comments
Changes in include fixing the docstring from
`tensorize_vllm_model.py`, check unexpected modules for LoRA
adapter loading with Tensorizer, and moving
`serialize_and_deserialize_lora` to `tests/lora/test_llama_tp.py`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar commented on vllm/lora/models.py:
```diff
@@ -219,10 +219,23 @@ def from_local_checkpoint(
+            lora_tensor_path = os.path.join(tensorizer_config.tensorizer_dir,
+                                            "adapter_model.tensors")
+            tensorizer_args = tensorizer_config._construct_tensorizer_args()
+            tensors = TensorDeserializer(lora_tensor_path,
```
Yes. [Added.](https://github.com/vllm-project/vllm/pull/17926/commits/ff0c40858db781539ee8ce68fd3c25e10b17ddb7)

I would've liked to have `check_unexpected_modules` be called at some mutual point in the code path regardless of loading mechanism (Tensorizer or SafeTensors), but that would've required modifying the lifetime of the SafeTensors file-like object, so I've called it for each.

sangstar commented on tests/lora/test_lora_with_tensorizer.py:
```diff
@@ -0,0 +1,61 @@
+    ("model_name", "lora_uri"),
+    [("meta-llama/Llama-2-7b-hf", "yard1/llama-2-7b-sql-lora-test"),
+     ("baichuan-inc/Baichuan-7B", "jeeejeee/baichuan7b-text2sql-spider")])
+def test_serialize_and_deserialize_lora(tmp_path, model_name, lora_uri):
```
[Done here.](https://github.com/vllm-project/vllm/pull/17926/commits/ff0c40858db781539ee8ce68fd3c25e10b17ddb7)

sangstar committed:
chore: Fix linter error
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

jeejeelee commented on tests/lora/test_llama_tp.py:
```diff
@@ -153,3 +184,41 @@ def test_llama_lora_tp4_fully_sharded_loras(sql_lora_files):
+
+    from vllm import LLM, EngineArgs
+    from vllm.model_executor.model_loader.tensorizer import (
+        TensorizerConfig, tensorize_lora_adapter, tensorize_vllm_model)
```
I'm curious why lazy import is used?

sangstar replied:
This was a defensive decision to avoid any circular imports I encountered when developing, although I'm no longer able to reproduce this. I've moved the import statements in this PR to be loaded eagerly with the rest.

jeejeelee commented on vllm/lora/peft_helper.py:
```diff
@@ -90,11 +90,30 @@ def from_dict(cls, config_dict: dict) -> "PEFTHelper":
     def from_local_dir(cls, lora_path: str,
-                       max_position_embeddings: Optional[int]) -> "PEFTHelper":
+                       max_position_embeddings: Optional[int],
+                       tensorizer_config_dict: Optional[dict]) -> "PEFTHelper":
```
```suggestion
                       tensorizer_config_dict: Optional[dict] = None) -> "PEFTHelper":
```

jeejeelee commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1449,6 +1449,16 @@ def load_model(self) -> None:
+        self,
+        tensorizer_config,
+    ) -> None:
+        from vllm.model_executor.model_loader import TensorizerLoader
```
NIT: add type hint for `tensorizer_config`.
BTW: I've seen a lot of code using lazy import, can you explain the reason ?

sangstar replied:
Resolved. Replied to and resolved the stuff about lazy imports in a separate conversation.

sangstar committed:
Update vllm/lora/peft_helper.py
Signed-off-by: Sanger Steel <sangersteel@gmail.com>
Co-authored-by: Jee Jee Li <pandaleefree@gmail.com>

sangstar commented on tests/lora/test_llama_tp.py:
```diff
@@ -153,3 +184,41 @@ def test_llama_lora_tp4_fully_sharded_loras(sql_lora_files):
+        TensorizerConfig, tensorize_lora_adapter, tensorize_vllm_model)
+
+    model_ref = "meta-llama/Llama-2-7b-hf"
+    lora_path = "yard1/llama-2-7b-sql-lora-test"
```
Will do! Although `lora_path`'s appropriate value here for this test is `sql_lora_huggingface_id`, so will do so.

sangstar committed:
fix: Resolve next round of review comments

Changes include eagerly importing Tensorizer modules,
applying type annotation fixes, fixing the Tensorizer
test in test_llama_tp.py, and fixing the test to
cleanup after serializing using a subprocess
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

jeejeelee reviewed (approved):
Overall LGTM, thank you for your contribution and paitence

sangstar committed:
fix: Get correct absolute path to example script
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Use `enforce_eager=True` for test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar commented:
@jeejeelee Fixed a bug in the LoRA test related to not properly getting the root dir, fixed using `VLLM_PATH`. I also encountered a weird bug for my test `tests/lora/test_llama_tp.py::test_serialize_and_deserialize_lora`, even though the test passes:

```
  File "/vllm-workspace/vllm/config.py", line 4494, in __str__
    f"compilation_config={self.compilation_config!r}")
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vllm-workspace/vllm/config.py", line 3875, in __repr__
    json.dumps(v)
  File "/usr/lib/python3.12/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type PostGradPassManager is not JSON serializable
```

This looks to be due to the fact that the `inductor_compile_config` field for `CompilationConfig` has a possible key `post_grad_custom_post_pass` with value type `PostGradPassManager` that is not JSON-serializable. Is this a known bug?

sangstar committed:
tests: Use tp=2 for LoRA tensorizer test

Uses tp = 2 to get this test
passing on L4s, while also
unit testing sharded
tensorizer support with LoRA.
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Add `@multi_gpu_test` decorator for tp=2 test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar commented:
@jeejeelee Added a tweak to get my LoRA test passing on the L4s on the test runner. Tests are passing except for a couple which look unrelated to my changes.

sangstar commented:
Rebased. Looks like some entrypoint tests had some failures but they're unrelated to my tests or changes.

jeejeelee commented:
@DarkLight1337  Can we force merge?
